### ISSUE TITLE
Inheritance 

### ISSUE THREAD

**Issue by serjooo** [sockeqwe/mosby-conductor#10]
So in conductor we would usually be using ButterKnife, and the way conductor handled the lifecycle of butterknife is by creating a class ButterKnifeController and then creating a class BaseController extending the ButterKnifeController, and then your own customized Controllers can extend the BaseController, and have all the basics handled for them all. 

Now mosby + conductor, what you can do is allow your ButterKnifeController also extend the MvpController and you get MvpController functionality. However, the issue arises when you want to actually use the MvpViewStateController instead of the basic MvpController to handle states. How would you go about extending this chain so you can actually pass the Generic Types of <View, Presenter, ViewState> while extending that Base Controller.

In short:

`public abstract class BaseController extends ButterKnifeController {

}

public abstract class ButterKnifeController extends MvpViewStateController {

}

public class HomeController extends BaseController<View, Presenter, ViewState>{
     //How would we go about passing these generics as you are able to do it
    //When you extend directly from MvpViewStateController<HomeView, HomePresenter, HomeViewState>
}`

Thank you for your time in advance.

**Comment by sockeqwe** [comment]
Hi, 
Mosby follows the principle of delegation over inheritance. Mosby offers a `*Delegate` class (which is basically a lifecycle listener) and the corresponding `*DelegateCallback` (an interface your controller has to implement with methods like `createPresenter()` , createViewState() etc.).

For mvp only (no viewstate) take a look at this package (contains the delegate class and the callback interface):
https://github.com/sockeqwe/mosby-conductor/tree/master/mvp/src/main/java/com/hannesdorfmann/mosby/mvp/conductor/delegate

For ViewState take a look at this package (contains the delegate class and the callback interface):
https://github.com/sockeqwe/mosby-conductor/tree/master/viewstate/src/main/java/com/hannesdorfmann/mosby/conductor/viewstate/delegate



Finally your controller has to implement the callback interface and you have to register the delegate to your controller, like this:

https://github.com/sockeqwe/mosby-conductor/blob/master/mvp/src/main/java/com/hannesdorfmann/mosby/mvp/conductor/MvpController.java#L24


Hope that helps.

**Comment by serjooo** [comment]
Hey, 

Thank you for your quick reply very helpful, I'll look at how to connect things right now by delegation. By the way, thank you for creating the MVP library great work! Also, watched a workshop you gave in Russia about android architecture patterns great stuff. Good luck. Thank you again!

**Comment by serjooo** [comment]
From what I understood I came up with this. Just wondering if I have done it correctly (Sorry I'm not a pro developer trying to be).

`public abstract class ButterKnifeController extends MvpViewStateController {

    private Unbinder unbinder;


    protected ButterKnifeController() {

    }

    /**
     * Called when the view is required to be inflated.
     * @param inflater The LayoutInflater instance.
     * @param container The view container.
     * @return The inflated view.
     */
    protected abstract View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container);

    /**
     * Called when the view has successfully been bound.
     * @param view The view.
     */
    protected void onViewBound(@NonNull View view) { }

    /**
     * Called when the view is instantiated.
     * Binds ButterKnife to the view.
     * @param inflater LayoutInflater instance.
     * @param container The view container.
     * @return The inflated view.
     */
    @NonNull
    @Override
    protected View onCreateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        View view = inflateView(inflater, container);
        unbinder = ButterKnife.bind(this, view);
        onViewBound(view);
        return view;
    }

    /**
     * Called when the view is destroyed.
     * Un-binds ButterKnife from the view.
     * @param view The view to destroy.
     */
    @Override
    protected void onDestroyView(View view) {
        super.onDestroyView(view);
        unbinder.unbind();
    }



}`


Now this is my BaseController that each other Controller like LoginController, RegistrationController will extend to implement everything needed.

`public abstract class BaseController<V extends MvpView, P extends MvpPresenter<V>, VS extends ViewState<V>> extends ButterKnifeController implements MvpViewStateConductorDelegateCallback {

    protected P presenter;

    protected BaseController() { }

    // Initializer block
    {
        addLifecycleListener(getMosbyLifecycleListener());
    }

    /**
     * Returns the actionBar instance.
     * @return The actionBar instance.
     */
    protected ActionBar getActionBar() {
        ActionBarProvider actionBarProvider = (ActionBarProvider) getActivity();
        return actionBarProvider != null ? actionBarProvider.getSupportActionBar() : null;
    }

    @Override
    protected View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        return null;
    }

    @Override
    protected void onAttach(@NonNull View view) {
        setTitle();
        super.onAttach(view);
    }

    /**
     * Sets the actionbar title.
     */
    protected void setTitle() {
        Controller parentController = getParentController();
        while (parentController != null) {
            if (parentController instanceof BaseController && ((BaseController)parentController).getTitle() != null) {
                return;
            }
            parentController = parentController.getParentController();
        }

        String title = getTitle();
        ActionBar actionBar = getActionBar();
        if (title != null && actionBar != null) {
            actionBar.setTitle(title);
        }
    }

    /**
     * Gets the actionbar title.
     * Should be overridden by Controllers extending this
     * BaseController.
     * @return The action bar title.
     */
    protected String getTitle() {
        return null;
    }

    @Override
    public ViewState getViewState() {
        return null;
    }

    @Override
    public void setViewState(ViewState viewState) {

    }

    @NonNull
    @Override
    public ViewState createViewState() {
        return null;
    }

    @Override
    public boolean isRestoringViewState() {
        return true;
    }


}`


(not sure why the class names are not included in the code part I tried fixing it, but it wouldn't include it in the gray area)

Sorry if I'm asking how to do this and taking from your time. Thanks in advance.
